### ISSUE TITLE
Error while compiling file image/icc_4.ksy

### ISSUE THREAD

**Issue by sc07kvm** [kaitai-io/kaitai_struct_formats#347]
kaitai-struct-compiler -t python .\kaitai_struct_formats\image\icc_4.ksy

can't compare CalcArrayType(BytesLimitType(IntNum(1),None,false,None,None)) and CalcBytesType

**Comment by dgelessus** [comment]
Can you try if this still happens with version 0.9 (snapshot) of the compiler? This might have been fixed as part of kaitai-io/kaitai_struct#719.

**Comment by sc07kvm** [comment]
I tried on the latest version
```
kaitai-struct-compiler --version
kaitai-struct-compiler 0.9-SNAPSHOT
```

**Comment by generalmimon** [comment]
> Can you try if this still happens with version 0.9 (snapshot) of the compiler? 

> I tried on the latest version

Yeah, I also tried it in the [devel Web IDE](https://ide.kaitai.io/devel/), and the problem is in this situation:

https://github.com/kaitai-io/kaitai_struct_formats/blob/d1e58e36f26ecff1375c6022dc040342914ee4d1/image/icc_4.ksy#L1142-L1145

As you can see in the snippet above, the field `padding` (among others across the spec) uses `contents` + `repeat`. However, judging by the error message _(`can't compare CalcArrayType(BytesLimitType(IntNum(1),None,false,None,None)) and CalcBytesType`)_, the `contents` key is interpreted as the expected value of **the whole array**, not of the individual elements. The entire array stored in attribute `padding` has indeed the type `CalcArrayType(BytesLimitType(IntNum(1),None,false,None,None))`, which can be better shown on this equivalent snippet rewritten with the `valid` key (new in 0.9 - see https://github.com/kaitai-io/kaitai_struct/issues/435#issue-323332353):

```ksy
- id: padding
  size: 1
  valid:
    eq: [0x00]
```

However, the [`icc_4.ksy` spec](https://github.com/kaitai-io/kaitai_struct_formats/blob/d1e58e36f26ecff1375c6022dc040342914ee4d1/image/icc_4.ksy) compiles fine **in the stable 0.8 version** of the compiler (and thus also in the [stable Web IDE](https://ide.kaitai.io/)), and the above excerpt compiles to this:

```javascript
          Colorant.prototype._read = function() {
            //  ... 
            this.padding = new Array((32 - this.name.length));
            for (var i = 0; i < (32 - this.name.length); i++) {
              this.padding = this._io.ensureFixedContents([0]);
            }
```

In short: `contents` affected _individual elements_ in 0.8, but `valid` introduced in 0.9 applies to the _entire array_, and this has also silently changed the behavior of `contents`, because it has been internally unified with `valid` key inside the compiler.

But I don't think this BC break is intentional: it's just that the repetitions are often forgotten/overlooked when implementing a new feature in KS, because sometimes their support requires additional hard work (see what needed to be done for the `process` key - https://github.com/kaitai-io/kaitai_struct_compiler/pull/198).

So let's add a test into our [test suite](https://github.com/kaitai-io/kaitai_struct_tests) and make it work again in the KSC, the [`icc_4.ksy` spec](https://github.com/kaitai-io/kaitai_struct_formats/blob/d1e58e36f26ecff1375c6022dc040342914ee4d1/image/icc_4.ksy) is completely fine.

**Comment by Artoria2e5** [comment]
The current web IDE is failing too. I guess 0.9 has become stable?
